test/module/libs/result_test.cpp

```diff
@@ -0,0 +1,212 @@
+}
+
+using R = Result<int, std::string>;
+R increment_but_less_5(int a) {
+  if (a < 5) {
+    return Ok(a + 1);
+  } else {
+    return Error(":("s);
+  }
+}
+
+TEST(Result, monadic) {
+  R a = 1;
+  ASSERT_TRUE(a);
+
+  R b = a | increment_but_less_5 | increment_but_less_5;
+
+  ASSERT_TRUE(a);
+  ASSERT_TRUE(b);
+  ASSERT_EQ(*b, *a + 2);
+
+  R c = b | increment_but_less_5 | increment_but_less_5 | increment_but_less_5;
+  ASSERT_FALSE(c);
+  ASSERT_EQ(c.error(), ":("s);
+
+  // monadic sequence test
+  using Rm = Result<bool, bool>;
+  auto succeed = [](auto &&) -> Rm {
+    SUCCEED();
+    return Ok(true);
+  };
+
+  auto make_error = [](auto &&) -> Rm { return Error(false); };
+
+  auto fail = [](auto &&) -> Rm { ADD_FAILURE(); };
+
+  // 3 consecutive success operations which return value, then 4th returns
+  // error. if 5th is executed, than test should fail.
+  Rm test_ = Ok(true);
+  Rm r_ = test_ | succeed | succeed | succeed | make_error | fail;
+}
+
+TEST(Result, pattern_matching) {
+  R a = 1;
+  ASSERT_TRUE(a);
+
+  a.match([](const Ok_t<int> &v) { SUCCEED(); },
+          [](const Error_t<std::string> &e) { FAIL(); });
+
+  // create own visitor
+  auto ok_handler = [](const R::OkType &v) { SUCCEED(); };
+  auto error_handler = [](const R::ErrorType &v) { FAIL(); };
+  auto visitor = make_visitor(ok_handler, error_handler);
+  a.match(visitor);
+}
+
+TEST(Result, constant_result) {
+  const R a = Error("test"s);
+  ASSERT_EQ("test"s, a.error());
+}
+
+TEST(Result, accessors_constructors) {
+  using R = Result<std::string, std::string>;
+  R nonc = "ok"s;
+  ASSERT_TRUE(nonc) << "contains value, but false";
+  ASSERT_EQ(*nonc, "ok"s);
+  ASSERT_EQ(nonc->size(), (*nonc).size());
+
+  const R c = "ok"s;
+  ASSERT_TRUE(c) << "contains value, but false";
+
+  auto ok = Ok("ok"s);
+  R a1 = ok;
+  R a2{ok};
+  R a3(ok);
+  R a4(std::move(ok));
+  R a5(R(Ok("ok"s)));
+  R a6 = Error("err"s);
+  a6 = std::move(Ok("ok"s));
+
+  const auto V = Ok("ok"s);
+  R a7 = Error("err"s);
+  a7 = V;
+  ASSERT_TRUE(a1);
+  ASSERT_TRUE(a2);
+  ASSERT_TRUE(a3);
+  ASSERT_TRUE(a4);
+  ASSERT_TRUE(a5);
+  ASSERT_TRUE(a6);
+  ASSERT_EQ(*a1, "ok"s);
+
+  auto err = Error("err"s);
+  R b1 = err;
+  R b2{err};
+  R b3(err);
+  R b4(std::move(err));
+  R b5(R(Error("err"s)));
+  R b6 = Ok("ok"s);
+  b6 = std::move(Error("err"s));
+
+  const auto E = Error("err"s);
+  R b7 = Ok("ok"s);
+  b7 = E;
+  ASSERT_FALSE(b1);
+  ASSERT_FALSE(b2);
+  ASSERT_FALSE(b3);
+  ASSERT_FALSE(b4);
+  ASSERT_FALSE(b5);
+  ASSERT_FALSE(b6);
+  ASSERT_FALSE(b7);
+
+  // rvalue operator * and ->
+  ASSERT_TRUE(R("ok"s));
+  ASSERT_EQ(*R("ok"s), "ok"s);
+  ASSERT_EQ(R("ok"s)->size(), "ok"s.size());
+
+  const R c1 = Ok("ok"s);
+  ASSERT_FALSE(!c1);
+  ASSERT_EQ(c1->size(), (*c1).size());
+
+  // copy constructors
+  R v1 = Ok("okay"s);
+  R v2 = v1;
+  R v3 = R(Ok("okay"s));
+  R v4((R(Error("s"s))));
+  ASSERT_TRUE(v1);
+  ASSERT_TRUE(v2);
+  ASSERT_TRUE(v3);
+  ASSERT_FALSE(v4);
+}
```

Comment:
Also, I expected to look on mating test for same types, like for `Result<int, int>`.